pkg/fqdn/dnsproxy/shared_client.go

```diff
@@ -0,0 +1,357 @@
+type SharedClient struct {
+	serverAddr string
+
+	*dns.Client
+
+	// requests is closed when the client needs to exit
+	requests chan request
+	// wg is waited on for the client finish exiting
+	wg sync.WaitGroup
+
+	lock.Mutex // protects the fields below
+	refcount   int
+	conn       *dns.Conn
+}
+
+func newSharedClient(conf *dns.Client, serverAddr string) *SharedClient {
+	return &SharedClient{
+		refcount:   1,
+		serverAddr: serverAddr,
+		Client:     conf,
+		requests:   make(chan request),
+	}
+}
+
+// ExchangeShared dials a connection to the server on first invocation, and starts a handler
+// goroutines to send and receive responses, distributing them to appropriate concurrent caller
+// based on the DNS message Id.
+func (c *SharedClient) ExchangeShared(m *dns.Msg) (r *dns.Msg, rtt time.Duration, err error) {
+	return c.ExchangeSharedContext(context.Background(), m)
+}
+
+// handler is started when the connection is dialed
+func handler(wg *sync.WaitGroup, client *dns.Client, conn *dns.Conn, requests chan request) {
+	defer wg.Done()
+
+	responses := make(chan sharedClientResponse)
+
+	// receiverTrigger is used to wake up the receive loop after request(s) have been sent. It
+	// must be buffered to be able to send a trigger while the receive loop is not yet ready to
+	// receive the trigger, as we do not want to stall the sender when the receiver is blocking
+	// on the read operation.
+	receiverTrigger := make(chan struct{}, 1)
+	triggerReceiver := func() {
+		select {
+		case receiverTrigger <- struct{}{}:
+		default:
+		}
+	}
+
+	// Receive loop
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(responses)
+
+		// No point trying to receive until the first request has been successfully sent, so
+		// wait for a trigger first. receiverTrigger is buffered, so this is safe
+		// to do, even if the sender sends the trigger before we are ready to receive here.
+		<-receiverTrigger
+
+		for {
+			// This will block but eventually return an i/o timeout, as we always set
+			// the timeouts before sending anything
+			r, err := conn.ReadMsg()
+			if err == nil {
+				responses <- sharedClientResponse{r, 0, nil}
+				continue // receive immediately again
+			}
+
+			// handler is not reading on the channel after closing.
+			// UDP connections return net.ErrClosed, while TCP/TLS connections are read
+			// via the io package, which return io.EOF.
+			if errors.Is(err, net.ErrClosed) || errors.Is(err, io.EOF) {
+				return
+			}
+
+			// send error response to cancel all current requests.
+			responses <- sharedClientResponse{nil, 0, err}
+
+			// wait for a trigger from the handler after any errors. Re-reading in
+			// this condition could busy loop, e.g., if a read timeout occurred.
+			// receiverTrigger is buffered so that we catch the trigger that may
+			// have been sent while we sent the error response above.
+			_, ok := <-receiverTrigger
```

Comment:
why don't this be moved to top of for loop and remove the one outside for loop?

Reply:
That would not be equivalent: we'd block on `receiverTrigger` in every loop iteration, whereas we currently can hit the `continue` case which does not involve the `receiverTrigger`

Reply:
yep, sorry my bad. didn't see continue case